src/pages/blog/post/der-komplette-guide-rund-um-sql-injections.mdx

```diff
@@ -0,0 +1,303 @@
+Klassische Patterns wie
+
+```sql
+... WHERE page_id = 5 OR 1 = 1
+```
+
+werden problemlos erkannt.
+
+Dies bietet keinen ausreichenden Schutz gegen SQL-Injections, macht es dem Angreifer jedoch schwieriger. Dennoch kann dies als zusätzliche Maßnahme eingebunden werden.
+
+### Schulung der Mitarbeiter
+
+Upskilling ist immer eine gute Sache.
+Auch im Fall von SQL-Injections (oder Web-Security im Allgemeinen) können Schulungen und Trainings wahre Wunder wirken.
+
+Inhalt des Trainings könnte u.a. sein:
+Sichere Validierungs- und Programmiere- Praktiken
+Sicherheitsbewusstsein
+Verfahren zur Reaktion auf Vorfälle
+Datenbank-Konfiguration
+
+### Security Testing
+
+Scan doch mal die Webanwendung und die zugrunde liegende Datenbank regelmäßig auf Schwachstellen. Die Behebung oder Entschärfung erkannter Schwachstellen sollte entsprechende Priorität in der Entwicklung erhalten.
+
+Aber auch regelmäßig Sicherheitstests durch, z. B. Penetrationstests oder Code-Reviews, um potenzielle Schwachstellen zu ermitteln und die Wirksamkeit der vorhandenen Sicherheitskontrollen zu überprüfen, können hilfreich sein.
+
+## Kann ich meine Applikation automatisch auf SQL-Injections testen?
+
+Komplett ohne manuellen Aufwand ist es nicht möglich. Dennoch gibt es verschiedene Varianten, die dich bei der Absicherung unterstützen können:
+
+### Statische Analyse
+
+Die klassische statische Code-Analyse ist nicht auf SQL-Injections im Speziellen ausgelegt. Dennoch gibt es einige Tools, die entsprechende Unterstützung mitbringen. I.d.R. muss dennoch das Model bzw. entsprechende Tool auf die jeweilige Programmiersprache und sogar auf die verwendete Library angepasst werden. 
+
+Ein positiver Seiteneffekt: Diese Tools können auch andere Code-Qualitätsprobleme aufzeigen, die zu Sicherheitslücken führen könnten.
+
+### Dynamische Analyse
+
+Tools zur dynamischen Code-Analyse simulieren Benutzereingaben und können dadurch Injektionspunkte identifizieren.
+
+Der Platzhirsch in diesem Bereich ist [sqlmap](https://sqlmap.org/) mit support für fast alle SQL basierten Datenbanken, inklusive MySQL, MSSQL, Oracle, PostgreSQL, Amazon Redshift, CockroachDB und Apache Ignite sowie den SQL-Injection-Arten wie Boolean-Based Blind, Time-Based Blind, Error-Based, UNION Query-Based, Stacked Queries und Out-of-Band.
+
+Eine Alternative bietet [OWASP ZAP](https://www.zaproxy.org/). Nicht speziell für SQL Injections ausgelegt, bietet dennoch Unterstützung für diesen Angriffsvektor.
+
+### Fuzzing
+
+Aktuell gibt es Research im Bereich Fuzzing und wie man mit Hilfe von Machine Learning SQL Injections aufspüren kann. Das entsprechende Paper [“SQLi-Fuzzer: A SQL Injection Vulnerability Discovery Framework Based on Machine Learning”](https://ieeexplore.ieee.org/document/9657925) wurde im Jahr 2021 veröffentlicht.
+Diverse proprietäre Software-Services werben bereits damit. Im Open Source Bereich ist (unseres Wissens nach) noch nichts fertig Benutzbares veröffentlicht worden.
```

Comment:
noch nichts out-of-the-box einsetzbares veröffentlicht

Reply:
```suggestion
Diverse proprietäre Software-Services werben bereits damit. Im Open Source Bereich ist (unseres Wissens nach) noch nichts Out-of-the-Box einsetzbares veröffentlicht worden.
```